components/wifi/esp8266-driver/ESP8266/ESP8266.cpp

```diff
@@ -576,31 +577,50 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
     _smutex.lock();
     set_timeout(ESP8266_SEND_TIMEOUT);
     _busy = false;
-    if (_parser.send("AT+CIPSEND=%d,%lu", id, amount)
-            && _parser.recv(">")
-            && _parser.write((char *)data, (int)amount) >= 0
-            && _parser.recv("SEND OK")) {
-        // No flow control, data overrun is possible
-        if (_serial_rts == NC) {
-            while (_parser.process_oob()); // Drain USART receive register
-        }
-        _smutex.unlock();
-        return NSAPI_ERROR_OK;
+    _error = false;
+    if (!_parser.send("AT+CIPSEND=%d,%lu", id, amount)) {
+        tr_debug("ESP8266::send(): AT+CIPSEND failed");
+        goto END;
     }
-    if (_error) {
-        _error = false;
+
+    if(!_parser.recv(">")) {
+        tr_debug("ESP8266::send(): didn't get \">\"");
+        ret = NSAPI_ERROR_WOULD_BLOCK;
+        goto END;
     }
+
+    if (_parser.write((char *)data, (int)amount) >= 0 && _parser.recv("SEND OK")) {
+        ret = NSAPI_ERROR_OK;
+    }
+
+END:
+    _process_oob(ESP8266_RECV_TIMEOUT, true); // Drain USART receive register to avoid data overrun
+
+    // error hierarchy, from low to high
     if (_busy) {
-        set_timeout();
-        _smutex.unlock();
-        tr_debug("returning WOULD_BLOCK");
-        return NSAPI_ERROR_WOULD_BLOCK;
+        ret = NSAPI_ERROR_WOULD_BLOCK;
+        tr_debug("ESP8266::send(): modem busy");
+    }
+
+    if (ret == NSAPI_ERROR_DEVICE_ERROR) {
+        ret = NSAPI_ERROR_WOULD_BLOCK;
+        tr_debug("ESP8266::send(): send failed");
+    }
+
+    if (_error) {
+        ret = NSAPI_ERROR_CONNECTION_LOST;
+        tr_debug("ESP8266::send(): connection disrupted");
```

Comment:
is this actually true?

Should we test that socket is actually closed? We should know that, because OOBs are already handled.

Reply:
> Should we test that socket is actually closed?
That is done on the next if-clause.

> If the connection cannot be established or gets
disrupted during data transmission, the system
returns:
ERROR

From the documentation.

Reply:
Ah.. true.. Thanks.